crates/loro-internal/src/handler.rs

```diff
@@ -158,95 +194,621 @@ impl BasicHandler {
+    inner: MaybeDetached<TreeInner>,
 }
 
-impl HandlerTrait for TreeHandler {
-    fn inner(&self) -> &BasicHandler {
-        &self.inner
+#[derive(Clone)]
+struct TreeInner {
+    next_counter: Counter,
+    map: FxHashMap<TreeID, MapHandler>,
+    parent_links: FxHashMap<TreeID, Option<TreeID>>,
+}
+
+impl TreeInner {
+    fn new() -> Self {
+        TreeInner {
+            next_counter: 0,
+            map: FxHashMap::default(),
+            parent_links: FxHashMap::default(),
+        }
+    }
+
+    fn create(&mut self, parent: Option<TreeID>) -> TreeID {
+        let id = TreeID::new(PeerID::MAX, self.next_counter);
+        self.next_counter += 1;
+        self.map.insert(
+            id,
+            Handler::new_unattached(ContainerType::Map)
+                .into_map()
+                .unwrap(),
+        );
+        self.parent_links.insert(id, parent);
+        id
+    }
+
+    fn delete(&mut self, id: TreeID) {
+        self.map.remove(&id);
+        self.parent_links.remove(&id);
+    }
+
+    fn get_parent(&self, id: TreeID) -> Option<Option<TreeID>> {
+        self.parent_links.get(&id).cloned()
+    }
+
+    fn mov(&mut self, target: TreeID, new_parent: Option<TreeID>) {
+        let old = self.parent_links.insert(target, new_parent);
+        assert!(old.is_some());
+    }
+
+    fn get_children(&self, id: TreeID) -> Option<Vec<TreeID>> {
+        let mut children = Vec::new();
+        for (c, p) in &self.parent_links {
+            if p.as_ref() == Some(&id) {
+                children.push(*c);
+            }
+        }
+        Some(children)
+    }
+}
+
+impl HandlerTrait for TreeHandler {
+    fn to_handler(&self) -> Handler {
+        Handler::Tree(self.clone())
+    }
+
+    fn attach(
+        &self,
+        _txn: &mut Transaction,
+        parent: &BasicHandler,
+        self_id: ContainerID,
+    ) -> LoroResult<Self> {
+        match &self.inner {
+            MaybeDetached::Dettached(t) => {
+                let mut t = t.try_lock().unwrap();
+                let inner = create_handler(parent, self_id);
+                let tree = inner.into_tree().unwrap();
+                if t.value.map.is_empty() {
+                    t.attached = tree.attached_handler().cloned();
+                    Ok(tree)
+                } else {
+                    unimplemented!("attach detached tree");
+                }
+            }
+            MaybeDetached::Attached(_a) => unreachable!(),
+        }
+    }
+
+    fn is_attached(&self) -> bool {
+        self.inner.is_attached()
+    }
+
+    fn attached_handler(&self) -> Option<&BasicHandler> {
+        self.inner.attached_handler()
+    }
+
+    fn get_value(&self) -> LoroValue {
+        match &self.inner {
+            MaybeDetached::Dettached(a) => todo!(),
+            MaybeDetached::Attached(a) => a.get_value(),
+        }
+    }
+
+    fn get_deep_value(&self) -> LoroValue {
+        match &self.inner {
+            MaybeDetached::Dettached(a) => todo!(),
```

Comment:
todo

Reply:
Unattached containers for trees are left intentionally for another day. Its behaviors are complex to implement correctly.